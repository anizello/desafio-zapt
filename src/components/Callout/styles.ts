import styled, { css, DefaultTheme } from 'styled-components'
import media from 'styled-media-query'
import { CallOutProps } from '.'

type WrapperProps = Pick<CallOutProps, 'imageDirection'>

const wrapperModifiers = {
  left: (theme: DefaultTheme) => css`
    grid-template-areas: 'content';
    grid-template-columns: 1fr;

    ${media.greaterThan('medium')`
      grid-template-areas: 'image content';
      grid-template-columns: 1fr 2fr;
    `}

    ${Image} {
      border-bottom-left-radius: ${theme.border.radius};
      border-top-left-radius: ${theme.border.radius};

      mask-image: url('/img/layout/waves-right.svg');
    }

    ${ShapePrimary} {
      top: -1.5rem;
      left: -2.5rem;
    }

    ${ShapeSecondary} {
      bottom: -3rem;
      right: -4rem;
    }

    ${ShapeSemiCircle} {
      bottom: 4rem;
      left: 5%;
      transform: scaleX(-1);

      ${media.lessThan('large')`
        bottom: 6rem;
      `}
    }

    ${ShapeCircle} {
      top: 2rem;
      left: 20%;

      ${media.lessThan('large')`
        top: 4rem;
      `}
    }
  `,

  right: (theme: DefaultTheme) => css`
    grid-template-areas: 'content';
    grid-template-columns: 1fr;

    ${media.greaterThan('medium')`
      grid-template-areas: 'content image';
      grid-template-columns: 2fr 1fr;
    `}

    ${Image} {
      border-bottom-right-radius: ${theme.border.radius};
      border-top-right-radius: ${theme.border.radius};

      mask-image: url('/img/layout/waves-left.svg');
    }

    ${ShapePrimary} {
      top: -1.5rem;
      left: -2.5rem;
    }

    ${ShapeSecondary} {
      bottom: -3rem;
      right: -4rem;
    }

    ${ShapeSemiCircle} {
      bottom: 4rem;
      right: 5%;

      ${media.lessThan('large')`
        bottom: 6rem;
      `}
    }

    ${ShapeCircle} {
      top: 2rem;
      right: 20%;

      ${media.lessThan('large')`
        top: 4rem;
      `}
    }
  `
}

export const Wrapper = styled.section<WrapperProps>`
  ${({ theme, imageDirection }) => css`
    display: grid;
    position: relative;
    min-height: 30rem;

    background-color: ${theme.colors.backgroundGray};
    border-radius: ${theme.border.radius};

    ${!!imageDirection && wrapperModifiers[imageDirection](theme)}
  `}
`

export const ContentWrapper = styled.div`
  ${({ theme }) => css`
    grid-area: content;
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: 4rem 4rem;

    color: ${theme.colors.gray};
    z-index: ${theme.layers.base};

    ${media.greaterThan('small')`
      padding: 5rem 5rem;
    `}

    ${media.greaterThan('medium')`
      padding: 5.6rem 10rem;
    `}
  `}
`

export const Title = styled.h1`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-size: ${theme.font.sizes.xl};
    font-weight: ${theme.font.regular};
    line-height: 3.4rem;

    padding-bottom: ${theme.spacings.xs};
  `}
`

export const SubTitle = styled.h3`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.base};
    font-weight: ${theme.font.regular};
    line-height: 2.4rem;
  `}
`

export const Content = styled.div`
  ${({ theme }) => css`
    margin-top: ${theme.spacings.lg};
  `}
`

export const ImageWrapper = styled.div`
  position: relative;
  height: 100%;
`

export const Image = styled.img`
  grid-area: image;
  width: 100%;
  height: 100%;
  max-width: 41.8rem;

  object-fit: cover;
  mask-repeat: repeat-y;
  mask-size: contain;
  mask-position: top;
`

export const Details = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
`

const Shape = styled.img`
  position: absolute;
`

export const ShapePrimary = styled(Shape)`
  ${media.lessThan('medium')`
     width: 5rem;
  `}

  ${media.lessThan('small')`
     width: 4rem;
  `}
`

export const ShapeSecondary = styled(Shape)`
  ${media.lessThan('medium')`
     width: 9rem;
  `}

  ${media.lessThan('small')`
     width: 8rem;
  `}
`

export const ShapeSemiCircle = styled(Shape)`
  ${media.lessThan('large')`
     width: 13rem;
  `}
`

export const ShapeCircle = styled.span`
  width: 6rem;
  height: 6rem;
  display: block;
  position: absolute;

  background: white;
  border-radius: 100%;

  ${media.greaterThan('large')`
    width: 8rem;
    height: 8rem;
  `}
`
