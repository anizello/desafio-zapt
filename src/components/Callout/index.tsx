import MediaMatch from 'components/MediaMatch'

import * as S from './styles'

export type CallOutProps = {
  title?: string
  subtitle?: string
  children?: React.ReactNode
  image?: string
  imageDirection?: 'left' | 'right'
  imageText?: string
  details?: boolean
}

const Callout = ({
  title,
  subtitle,
  children,
  image,
  imageDirection = 'right',
  imageText,
  details = false
}: CallOutProps) => (
  <S.Wrapper imageDirection={imageDirection}>
    <S.ContentWrapper>
      {!!title && <S.Title>{title}</S.Title>}
      {!!subtitle && <S.SubTitle>{subtitle}</S.SubTitle>}
      {!!children && <S.Content>{children}</S.Content>}
    </S.ContentWrapper>

    <MediaMatch greaterThan="medium">
      <S.ImageWrapper>
        <S.Image src={image} alt={imageText} />
      </S.ImageWrapper>
    </MediaMatch>

    {!!details && (
      <S.Details>
        <S.ShapePrimary
          src="/img/layout/float-primary.svg"
          aria-label="Shape Primary"
        />
        <MediaMatch greaterThan="medium">
          <S.ShapeCircle aria-label="Shape Circle" />
          <S.ShapeSemiCircle
            src="/img/layout/semi-circle.svg"
            aria-label="Shape Semi Circle"
          />
        </MediaMatch>
        <S.ShapeSecondary
          src="/img/layout/float-secondary.svg"
          aria-label="Shape Secondary"
        />
      </S.Details>
    )}
  </S.Wrapper>
)

export default Callout
