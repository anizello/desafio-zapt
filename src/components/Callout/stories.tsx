import { Story, Meta } from '@storybook/react'
import Callout, { CallOutProps } from '.'
import Button from 'components/Button'

export default {
  title: 'Callout',
  component: Callout,
  args: {
    title: 'Bem vindo Márcio,',
    subtitle:
      'Para você aproveitar nossas oportunidades primeiro você precisa estar dentro de um grupo, estamos aqui para lhe ajudar. ',
    image: '/img/promo/callout.jpeg',
    imageDirection: 'right',
    imageText: 'Três mulheres abraçadas e sorrindo',
    hasDetails: false
  },
  argTypes: {
    imageText: {
      table: {
        disable: true
      }
    },
    children: {
      table: {
        disable: true
      }
    }
  }
} as Meta

export const WithDetails: Story<CallOutProps> = (args) => (
  <div style={{ maxWidth: '120rem', margin: '5rem auto' }}>
    <Callout {...args} />
  </div>
)

WithDetails.args = {
  hasDetails: true,
  children: <Button>Criar grupo</Button>
}

export const Default: Story<CallOutProps> = (args) => (
  <div style={{ maxWidth: '120rem', margin: '5rem auto' }}>
    <Callout {...args} />
  </div>
)

export const LeftImage: Story<CallOutProps> = (args) => (
  <div style={{ maxWidth: '120rem', margin: '5rem auto' }}>
    <Callout {...args} />
  </div>
)

LeftImage.args = {
  imageDirection: 'left'
}
