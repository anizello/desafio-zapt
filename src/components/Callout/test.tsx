import { render, screen } from 'utils/test-utils'

import Callout from '.'

const props = {
  title: 'Bem vindo Márcio,',
  subtitle:
    'Para você aproveitar nossas oportunidades primeiro você precisa estar dentro de um grupo, estamos aqui para lhe ajudar. ',
  image: '/img/promo/callout.jpeg',
  imageText: 'Três mulheres abraçadas e sorrindo'
}

describe('<Callout />', () => {
  it('should render the callout with default values', () => {
    const { container } = render(<Callout {...props} />)

    expect(container.firstChild).toHaveStyleRule(
      'grid-template-areas',
      "'content image'",
      {
        media: '(min-width: 768px)'
      }
    )

    expect(
      screen.getByRole('heading', { name: /bem vindo márcio,/i })
    ).toBeInTheDocument()

    expect(
      screen.getByRole('heading', {
        name: /para você aproveitar nossas oportunidades primeiro você precisa estar dentro de um grupo, estamos aqui para lhe ajudar./i
      })
    ).toBeInTheDocument()

    expect(
      screen.getByAltText(/três mulheres abraçadas e sorrindo/i)
    ).toHaveAttribute('src', '/img/promo/callout.jpeg')

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render callout with image on the left', () => {
    const { container } = render(<Callout {...props} imageDirection="left" />)

    expect(container.firstChild).toHaveStyleRule(
      'grid-template-areas',
      "'image content'",
      {
        media: '(min-width: 768px)'
      }
    )
  })

  it('should render with children', () => {
    render(
      <Callout {...props}>
        <span>A Children</span>
      </Callout>
    )

    expect(screen.getByText(/a children/i)).toBeInTheDocument()
  })

  it('should render with details', () => {
    render(<Callout {...props} details />)

    expect(screen.getByLabelText(/shape primary/i)).toBeInTheDocument()
    expect(screen.getByLabelText(/shape secondary/i)).toBeInTheDocument()
    expect(screen.getByLabelText(/shape circle/i)).toBeInTheDocument()
    expect(screen.getByLabelText(/shape semi circle/i)).toBeInTheDocument()
  })
})
