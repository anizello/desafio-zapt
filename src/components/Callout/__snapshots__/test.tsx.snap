// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Callout /> should render the callout with default values 1`] = `
.c4 {
  display: none;
}

.c0 {
  display: grid;
  position: relative;
  min-height: 30rem;
  background-color: #fafafa;
  border-radius: 0.4rem;
  grid-template-areas: 'content';
  grid-template-columns: 1fr;
}

.c0 .c6 {
  border-bottom-right-radius: 0.4rem;
  border-top-right-radius: 0.4rem;
  -webkit-mask-image: url('/img/layout/waves-left.svg');
  mask-image: url('/img/layout/waves-left.svg');
}

.c1 {
  grid-area: content;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 4rem 4rem;
  color: #7a7a7a;
  z-index: 10;
}

.c2 {
  color: #000000;
  font-size: 2.8rem;
  font-weight: 400;
  line-height: 3.4rem;
  padding-bottom: 0.8rem;
}

.c3 {
  font-size: 1.6rem;
  font-weight: 400;
  line-height: 2.4rem;
}

.c5 {
  position: relative;
  height: 100%;
}

.c7 {
  grid-area: image;
  width: 100%;
  height: 100%;
  max-width: 41.8rem;
  object-fit: cover;
  -webkit-mask-repeat: repeat-y;
  mask-repeat: repeat-y;
  -webkit-mask-size: contain;
  mask-size: contain;
  -webkit-mask-position: top;
  mask-position: top;
}

@media (min-width:768px) {
  .c4 {
    display: block;
  }
}

@media (min-width:768px) {
  .c0 {
    grid-template-areas: 'content image';
    grid-template-columns: 2fr 1fr;
  }
}

@media (max-width:1170px) {

}

@media (max-width:1170px) {

}

@media (min-width:450px) {
  .c1 {
    padding: 5rem 5rem;
  }
}

@media (min-width:768px) {
  .c1 {
    padding: 5.6rem 10rem;
  }
}

<section
  class="c0"
>
  <div
    class="c1"
  >
    <h1
      class="c2"
    >
      Bem vindo Márcio,
    </h1>
    <h3
      class="c3"
    >
      Para você aproveitar nossas oportunidades primeiro você precisa estar dentro de um grupo, estamos aqui para lhe ajudar. 
    </h3>
  </div>
  <div
    class="c4"
  >
    <div
      class="c5"
    >
      <img
        alt="Três mulheres abraçadas e sorrindo"
        class="c6 c7"
        src="/img/promo/callout.jpeg"
      />
    </div>
  </div>
</section>
`
